### ISSUE TITLE
Unified approach to colormapping

### ISSUE THREAD

**Issue by almarklein** [pygfx/pygfx#104]
Proposal:

### Colors 

In all cases, at some point you have a color that can consist of 1, 2, 3, 4 elements:
   * color is 1D: grayscale
   * color is 2D: gray + alpha
   * color is 3D: RGB
   * color is 4D: RGBA
   
### For meshes, lines and points

  * `geometry.texcoords` can be Nx1, Nx2, Nx3, Nx4, or omitted.
  * `material.map` dimension should match the texcoords.
  * If `material.map` is present, `geometry.texcoords` must be present too.
  * If `material.map` is not present, the `material.color` is used.

### For volumes and images

In this case there are two textures, one for the "data", and an optional one for the colormap.

   * If no colormap is suplied, the value of the volume/image is rendered as a color as describe above.
   * If a colormap is present, the data-values are used to sample into the colormap, and the `nchannes` of the data texture must match the dimensions of the colormap.

**Comment by almarklein** [comment]
#### Question about per-vertex coloring
A user can implement per-vertex coloring, by setting Nx3 texcoords (RGB values), and setting a "unit' colormap. But this is somewhat awkward, and involves an unnecessary texture lookup. Do we want `geometry.vertex_colors`? (for mesh, line, points).

#### Question about the data-texture

I feel that the texture containing the data is not "material"; its data intrinsic to the object, not a detail on how it should be rendered. I'd argue that its geometry (the size of the texture defines the geometry).

In other words, the ThreeJS abstraction, where everything is a mesh, does not hold up so well for our more exotic types of data. Perhaps we should broaden the scope of geometry to "data that defines an object" or something.

**Comment by Korijn** [comment]
### Per-vertex coloring

I think vertex coloring is such a common use case that we need to support it with a clear, simple API. I don't really see what would make it complicated? In threejs it's a flag on mesh materials just like wireframe rendering is. If `material.vertex_colors` is `True`, then `geometry.colors` is used.

**Comment by Korijn** [comment]
### On (volume) data-textures

I think it's important that we keep our object model consistent in a conceptual sense. It's a scene graph that tells the renderer what to draw (world objects, geometry), and how to draw it (materials, cameras). It's modelled to support that specific purpose, it's not modelled after the data.

You can wonder why we even have geometry objects separate from world objects; the main purpose of geometry objects is to enable structural sharing of data among world objects. That's really it's sole reason for existence. Similarly, I think we want to be able to do that with volume image data. We should be able to upload a data texture to the GPU once, and have a reference to it in our python object model, and share it among multiple world objects.

Given all of that, I think it makes a lot of sense to add the data texture to the geometry that is passed to our volume world objects. I think it will work just fine for our purposes!

---

I do want to point out that I also see users maybe wanting to put a quad somewhere in a scene, and render a slice from a 3d texture onto it. They should be able to reference the same data texture that also lives on the volume geometry, to avoid having to upload the same (potentially large) binary blob to the GPU twice.

**Comment by almarklein** [comment]
I don't think that I understand this example:

> Or, what if they want the MIP of a volume to be rendered on a box of a specific size and in a specific orientation? I think for those users it will be much more natural to put a BoxGeometry in a Mesh and give it a VolumeMaterial and pass it the data texture and use 3d texture coordinates.

If you want a MIP-rendered volume with a specific size and orientation, you'd have a "normal" volume use-case, for which you orient the object. If you want a cube with the result of a MIP shown as a texture, you should probably use a sub-scene approach with a 2nd renderer (that renders to the texture that is shown on the cube). I'm probably missing your point though :)
